File: java/src/main/java/lambda/ServiceOne.java
package lambda;

import saaf.Inspector;
import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.amazonaws.services.s3.model.ObjectMetadata;
import java.io.InputStreamReader;
import com.amazonaws.services.s3.AmazonS3ClientBuilder;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.model.S3Object;
import com.amazonaws.services.s3.model.GetObjectRequest;
import com.amazonaws.services.lambda.runtime.LambdaLogger;

import java.io.*;
import java.util.*;

/**
 * uwt.lambda_test::handleRequest
 *
 * @author Wes Lloyd
 * @author Robert Cordingly
 * @author David Perez
 */
public class ServiceOne implements RequestHandler<Request, HashMap<String, Object>> {

    public List<String[]> readCsv(InputStream input, LambdaLogger logger) throws IOException {
        // csvfile['Body'].read().decode('utf-8').split("\n") as input
        List<String[]> records = new ArrayList<String[]>();
        InputStreamReader isReader = new InputStreamReader(input);
        BufferedReader reader = new BufferedReader(isReader);
        String str;
        // change to for?
        while ((str = reader.readLine()) != null) {
            str = str + ",0,0";
            records.add(str.split(","));
            logger.log(str);
        }
        isReader.close();
        return records;
    }

    public StringBuilder writeCsv(List<String[]> records, LambdaLogger logger) {
        StringBuilder sb = new StringBuilder();

        records.get(0)[14]="Order Processing Time";
        records.get(0)[15]="Gross Margin";
        sb.append(String.join(",", records.get(0)) +"\n");
        Set<Integer> unique_ids = new HashSet<Integer>();

        try {
            for (int i = 1; i < records.size() - 1; i++) {
                if (unique_ids.contains(Integer.parseInt(records.get(i)[6]))) {
                    continue;
                } else {
                    String val = records.get(i)[4];
                    if (val.equals("C")) {
                        records.get(i)[4] = "Critical";
                    } else if (val.equals("L")) {
                        records.get(i)[4] = "Low";
                    } else if (val.equals("M")) {
                        records.get(i)[4] = "Medium";
                    } else if (val.equals("H")) {
                        records.get(i)[4] = "High";
                    }

                    String[] date1_values = (records.get(i)[5]).split("/");
                    String[] date2_values = (records.get(i)[7]).split("/");
                    int month = Integer.parseInt(date1_values[0]);
                    int day = Integer.parseInt(date1_values[1]);
                    int year = Integer.parseInt(date1_values[2]);

                    int month2 = Integer.parseInt(date2_values[0]);
                    int day2 = Integer.parseInt(date2_values[1]);
                    int year2 = Integer.parseInt(date2_values[2]);

                    int order_time = ((year2 - year) * 365) + ((month2 - month) * 30) + (day2 - day);
                    float gross_margin = Float.parseFloat((records.get(i)[13]))
                            / Float.parseFloat((records.get(i)[11]));
                    records.get(i)[14] = (Integer.toString(order_time));
                    records.get(i)[15] = (String.valueOf(gross_margin));
                    logger.log(records.get(i)[14]);
                    logger.log(records.get(i)[15]);
                    sb.append(String.join(",", records.get(i)) + "\n");
                    unique_ids.add(Integer.parseInt(records.get(i)[6]));
                }
            }
        } catch (Exception e) {
            throw new RuntimeException("Can't parse file " + e);
        }
        return sb;
    }

    public HashMap<String, Object> handleRequest(Request request, Context context) {
        Inspector inspector = new Inspector();
        inspector.inspectAll();

        inspector.addAttribute("message",
                "bucketname is = " + request.getBucketName() + "! This is an attributed added to the Inspector!");

        String bucketname = request.getBucketName();
        String key = request.getKey();
	LambdaLogger logger = context.getLogger();
	logger.log("start getobj");
        AmazonS3 s3Client = AmazonS3ClientBuilder.standard().build();
        S3Object s3Object = s3Client.getObject(new GetObjectRequest(bucketname, key));
        InputStream objectData = s3Object.getObjectContent();
	logger.log("finished getobj");
        StringBuilder sw = null;
        try {
            logger.log("started reading csv");
            List<String[]> records = readCsv(objectData, logger);
            logger.log("finished reading csv, starting to write csv");
            sw = writeCsv(records, logger);
            logger.log("finished writing csv");
            objectData.close();

        } catch (IOException e) {
            System.out.println("Error reading csv: " + e);
        }
        logger.log("turning sw to bytes");
        String fullstring=sw.toString();
        sw.setLength(0);
        byte[] bytes = fullstring.getBytes();
        logger.log("bytes array created");

        InputStream is = new ByteArrayInputStream(bytes);
        logger.log("inputstream created");
        ObjectMetadata meta = new ObjectMetadata();
        logger.log("out");
        meta.setContentLength(bytes.length);
        meta.setContentType("text/plain");
                logger.log("out2");

        AmazonS3 s3Client2 = AmazonS3ClientBuilder.standard().build();
        logger.log("out3");

        s3Client2.putObject(bucketname, key.substring(0, key.lastIndexOf('.')) + "/" + System.currentTimeMillis() + "_"
                + inspector.getAttribute("uuid") + ".csv", is, meta);
        logger.log("out4");

        inspector.addAttribute("value", "Bucket: " + bucketname + " key:" + key + " processed. record 0 = ");
        logger.log("out5");

        inspector.inspectAllDeltas();
        return inspector.finish();
    }
}

File: nodejs/src/serviceOne/serviceOne.js
const Inspector = require('../Inspector');
const RequestHandler = require("../requestHandler");
const CsvReader = require('../common/csvReader');
const CsvWriter = require('./csvWriter');

/**
 * @typedef ServiceOneRequest
 * @property {String} bucketName
 * @property {String} key
 */

/**
 * @typedef {function(): number} TimeTagProducer
 */

/**
 * @typedef {function(Inspector): *} UUIDProvider
 */

/**
 * @extends RequestHandler
 */
class ServiceOne extends RequestHandler {

    /**
     * @type {function(String):void}
     */
    _logger;

    /**
     * @type {S3}
     */
    _s3;

    /**
     * @type {TimeTagProducer}
     */
    _timeTagProducer;

    /**
     * @type {UUIDProvider}
     */
    _uuidProvider;

    /**
     * @type {CsvReader}
     */
    _csvReader;

    /**
     * @type {CsvWriter}
     */
    _csvWriter;

    /**
     * @param {function(string):void} logger
     * @param {S3} s3
     * @param {TimeTagProducer} timeTagProducer
     * @param {UUIDProvider} uuidProvider
     */
    constructor(logger, s3, timeTagProducer, uuidProvider) {
        super();
        this._logger = logger;
        this._s3 = s3;
        this._timeTagProducer = timeTagProducer;
        this._uuidProvider = uuidProvider;
        this._csvReader = new CsvReader(logger);
        this._csvWriter = new CsvWriter(logger);
    }

    /**
     * @param {ServiceOneRequest} request
     * @param context
     * @return {Promise<Attributes>}
     */
    async handleRequest(request, context) {
        let inspector = new Inspector();
        inspector.inspectAll();

        inspector.addAttribute("message", `bucketname is = ${request['bucketName']}! This is an attributed added to the Inspector!`);

        let body;
        try {
            body = await this._readRequestedFile(request);
        } catch (e) {
            this._log("Failed to read requested file");
            inspector.inspectAllDeltas();
            return inspector.finish();
        }
        let output = await this._produceCsvOutput(body);
        await this._uploadCsv(request, inspector, output);

        inspector.addAttribute("value", "Bucket: " + request.bucketName + " key:" + request.key + " processed. record 0 = ");
        inspector.inspectAllDeltas();

        return inspector.finish();
    }

    /**
     * @param {String} message
     * @private
     */
    _log(message) {
        this._logger.call(null, message);
    }

    /**
     * @param {ServiceOneRequest} request
     * @return {Promise<S3.Body>}
     */
    async _readRequestedFile(request) {
        this._log("start getobj");
        let s3Object = await this._s3.getObject({
            Bucket: request.bucketName,
            Key: request.key
        }).promise();

        /**
         * @type {S3.Body}
         */
        let body = s3Object.Body;
        this._log("finished getobj");
        return body;
    }

    /**
     * @param {S3.Body} body
     * @return {Promise<Buffer>}
     */
    async _produceCsvOutput(body) {
        let output;
        try {
            this._log("started reading csv");
            let records = await this._csvReader.read(body);
            this._log("finished reading csv, starting to write csv");
            output = this._csvWriter.write(records);
            this._log("finished writing csv");
        } catch (e) {
            this._log("Error reading csv: " + (typeof e === 'string' ? e : e.message));
        }
        return Buffer.from(output);
    }

    /**
     * @param {ServiceOneRequest} request
     * @param {Inspector} inspector
     * @param {Buffer} csv
     * @return {Promise<*>}
     */
    async _uploadCsv(request, inspector, csv) {
        let targetKey = `${request.key.substring(0, request.key.lastIndexOf('.'))}/${this._timeTagProducer.call()}_${this._uuidProvider.call(null, inspector)}.csv`;
        return this._s3.putObject({
            Bucket: request.bucketName,
            Key: targetKey,
            Body: csv
        }).promise();
    }
}

module.exports = ServiceOne;